password_security/models/res_users.py

```diff
@@ -0,0 +1,208 @@
+        data.update(
+            {
+                "password_lower": company_id.password_lower,
+                "password_upper": company_id.password_upper,
+                "password_numeric": company_id.password_numeric,
+                "password_special": company_id.password_special,
+                "password_length": company_id.password_length,
+                "password_estimate": company_id.password_estimate,
+            }
+        )
+        return data
+
+    def _check_password_policy(self, passwords):
+        result = super(ResUsers, self)._check_password_policy(passwords)
+
+        for password in passwords:
+            if not password:
+                continue
+            self._check_password(password)
+
+        return result
+
+    @api.model
+    def get_estimation(self, password):
+        return zxcvbn.zxcvbn(password)
+
+    def password_match_message(self):
+        self.ensure_one()
+        company_id = self.company_id
+        message = []
+        if company_id.password_lower:
+            message.append(
+                _("\n* Lowercase letter (at least %s characters)")
+                % str(company_id.password_lower)
+            )
+        if company_id.password_upper:
+            message.append(
+                _("\n* Uppercase letter (at least %s characters)")
+                % str(company_id.password_upper)
+            )
+        if company_id.password_numeric:
+            message.append(
+                _("\n* Numeric digit (at least %s characters)")
+                % str(company_id.password_numeric)
+            )
+        if company_id.password_special:
+            message.append(
+                _("\n* Special character (at least %s characters)")
+                % str(company_id.password_special)
+            )
+        if message:
+            message = [_("Must contain the following:")] + message
+        if company_id.password_length:
+            message = [
+                _("Password must be %d characters or more.")
+                % company_id.password_length
+            ] + message
+        return "\r".join(message)
+
+    def _check_password(self, password):
+        self._check_password_rules(password)
+        self._check_password_history(password)
+        return True
+
+    def _check_password_rules(self, password):
+        self.ensure_one()
+        if not password:
+            return True
+        company_id = self.company_id
+        password_regex = [
+            "^",
+            "(?=.*?[a-z]){" + str(company_id.password_lower) + ",}",
+            "(?=.*?[A-Z]){" + str(company_id.password_upper) + ",}",
+            "(?=.*?\\d){" + str(company_id.password_numeric) + ",}",
+            r"(?=.*?[\W_]){" + str(company_id.password_special) + ",}",
+            ".{%d,}$" % int(company_id.password_length),
+        ]
+        if not re.search("".join(password_regex), password):
+            raise UserError(self.password_match_message())
```

Comment:
Throw ValidationError here instead to match auth_saml test expectations?

Reply:
Alternately, #457 should work around this without having to change any exception classes.

Reply:
@amh-mw, I've done changes as per your feedback.